Reads from the CT Server (if set)

AndrewGorton opened the pull request:
Uses _&lt;register&gt;_.source to dictate if it uses PostGres or CTServer.

Commit:
Switched to SLF4J logger rather than our home-rolled one.

Commit:
Fetch data from the specified CTServer.

Commit:
With Phil's help, now creates the appropriate hash for use with get-proof-by-hash.

Commit:
Tests added to ensure far-end errors are propagated.

Commit:
Reinstated upload to CloudSearch ability.

Commit:
Removed the string comparison to see if we should use the CT server.

Commit:
Commented out display of the reconstructed payload

Commit:
Removed commented-out code

Commit:
Added some unit tests for the EntryParser

Commit:
Moved the signature generation outside of the EntryParser.

Commit:
Missing file added.

philandstuff commented on src/main/java/uk/gov/indexer/ctserver/Audit.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public class Audit {
```
I'm not sure Audit is the right name for this -- based on the [endpoint name & docs](http://tools.ietf.org/html/rfc6962#section-4.5), maybe "Proof" or "AuditProof"?

philandstuff commented on src/main/java/uk/gov/indexer/ctserver/CTServer.java:
```diff
@@ -0,0 +1,86 @@
+import java.security.MessageDigest;
+import java.util.Base64;
+
+public class CTServer {
```
I wonder if, longer term, we'll want to extract this into a client library? Not something for this PR anyway.

philandstuff replied:
Mentioned in slack but recording here for posterity: there already is a [java ct client](https://github.com/google/certificate-transparency/blob/master/java/src/org/certificatetransparency/ctlog/comm/HttpLogClient.java) but it's not clear how to build it with the gclient workflow and I don't think any jars are published for it.

EDIT: this is just for information, not something blocking the PR or anything.

Commit:
Renamed Audit to AuditProofs as per Phil's request

om-sharma commented on src/main/java/uk/gov/indexer/Indexer.java:
```diff
@@ -21,21 +28,29 @@ public Indexer(Configuration configuration, String register) {
-                    10,
-                    TimeUnit.SECONDS
-            );
+            if (configuration.getCTServerEndpointForRegister(register).isPresent()) {
```
can be simplified as below: 

           Fetcher fetcher = configuration.getCTServerEndpointForRegister(register)
                    .map(endPoint ->
                                    (Fetcher) new CTFetcher(new CTServer(endPoint))
                    ).orElse(
                            new PGFetcher(sourceDBQueryDAO)
                    );

            executorService.scheduleAtFixedRate(
                    new IndexerTask(register, fetcher, destinationDBUpdateDAO),
                    0,
                    10,
                    TimeUnit.SECONDS);

Commit:
Simplified the code as requested by Om

Commit:
Reverted changes requested by Om as temporary variable was needed.
Fixed unused imports.
Removed places where I'd created a temp variable immediately before return.

Commit:
Moved DBI creation to inside ifcheck for ctserver.

Commit:
Merge branch 'master' of github.com:openregister/indexer into read-from-ct-server

philandstuff commented on src/main/java/uk/gov/indexer/ctserver/AuditProofs.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public class AuditProofs {
```
nitpick: I think this is a single AuditProof, not a collection of many AuditProofs

philandstuff commented:
@om-sharma is there anything holding this back?

Commit:
Renamed to singular AuditProof as requested.

Commit:
Now hashes the raw JSON data, not the MerkleTreeLeaf.

Commit:
Converted item hash to be hex rather than Base64 encoding.

Commit:
Ensure we close the response object.

Commit:
Ensure we close the response object.

Commit:
Merge branch 'read-from-ct-server' of github.com-andrewgortonuk:openregister/indexer into read-from-ct-server

Commit:
When using the CTServer, a large batch size can cause Indexer to time out
waiting for a response.

Indexer can send newline-delimited lists for batch updates. If you're
using CTServer, each item is sent to CTServer in turn. The CTServer takes
a second to process each, causing Mint to not respond before Indexer stops
waiting for a response and throws a timeout error.
